### PR TITLE
MASISMIG-37628 - Add Cognos Analytics installation to the MAS Ansible Playbook

### PR DESCRIPTION
**Pull request by briangar-ibm**
MASISMIG-37628 - Add Cognos Analytics installation to the MAS Ansible Playbook

### CODE REVIEW

**Review by lokesh-sreedhara — APPROVED** [review]

**Comment by briangar-ibm** [comment]
> @briangar-ibm what about the CAService instance CR. will that be part of the ansible automation?

That will be a manual process.  The requirement to create a platform connection in CPD for both the Cognos Content Store and Audit function will be manual process for now.  We can look at adding this in at a later date if it is feasible to do so.

**Comment by lokesh-sreedhara** [comment]
there was a conflict on default.yml. jus resolved it

File: ibm/mas_devops/roles/cp4d_service/defaults/main.yml [hunk 1]
```diff
@@ -46,6 +46,7 @@ cpd_service_name: "{{ lookup('env', 'CPD_SERVICE_NAME') }}"
 # The storage class to use for this CP4D service
 cpd_service_storage_class: "{{ lookup('env', 'CPD_SERVICE_STORAGE_CLASS') }}"
 cpd_service_block_storage_class: "{{ lookup('env', 'CPD_SERVICE_BLOCK_STORAGE_CLASS') }}" # needed for WD
+cpd_service_file_storage_class: "{{ lookup('env', 'CPD_SERVICE_FILE_STORAGE_CLASS') }}" # needed for Cognos Analytics
```
Comment (andrercm) [inline ibm/mas_devops/roles/cp4d_service/defaults/main.yml:49]: will `cpd_service_file_storage_class ` value be any different from `cpd_service_storage_class ` ? Normally `cpd_service_storage_class` is already used primarily for file storage classes so this new var should not be needed.

File: ibm/mas_devops/roles/cp4d_service/tasks/determine-storage-classes.yml [hunk 2]
```diff
@@ -21,19 +21,37 @@
...
-- name: Default Metadata Storage if not set by user (Watson Discovery only)
+- name: Default Metadata Storage if not set by user (Watson Discovery)
   when:
     - cpd_service_name == "wd"
     - cpd_service_storage_class is not defined or cpd_service_storage_class == ""
     - cpd_service_block_storage_class is not defined or cpd_service_block_storage_class == ""
   set_fact:
     cpd_service_storage_class: "{{ (cpd_minor_version == '4.6') | ternary(lookup_storageclasses | ibm.mas_devops.defaultStorageClass(default_storage_classes_rwx), lookup_storageclasses | ibm.mas_devops.defaultStorageClass(default_storage_classes_rwo)) }}" # if cpd 4.6 then we need RWX as SC, otherwise it will be RWO.
     cpd_service_block_storage_class: "{{ lookup_storageclasses | ibm.mas_devops.defaultStorageClass(default_storage_classes_rwo) }}" # This var is only needed by WD
+- name: Default CPD Service Storage if not set by user (Cognos Analytics only)
```
Comment (andrercm) [inline ibm/mas_devops/roles/cp4d_service/tasks/determine-storage-classes.yml:37]: If Cognos uses only file storage class and does not need an specific file storage class then we probably dont need this code and we can live with the defaulted options? i.e ibmc-file-gold-gid ?

File: ibm/mas_devops/roles/cp4d_service/templates/services/ca.yml.j2 [hunk 3]
```diff
@@ -0,0 +1,15 @@
+apiVersion: ca.cpd.ibm.com/v1
+kind: "{{ cpd_service_info[cpd_service_name].crd_kind }}"
+metadata:
+  name: "{{ cpd_service_info[cpd_service_name].cr_name }}"
+  namespace: "{{ cpd_instance_namespace }}"
+spec:
+  blockStorageClass: "{{ cpd_service_block_storage_class }}"
+  fileStorageClass: "{{ cpd_service_file_storage_class }}"
```
Comment (andrercm) [inline ibm/mas_devops/roles/cp4d_service/templates/services/ca.yml.j2:8]: same as above comments, can we make it like other cpd roles that use file storage class and have just `cpd_service_storage_class ` as it's normally auto set to the defaulted file storage class for the targeted cloud provider?

**Review by andrercm — COMMENTED** [review]
Hi Brian, would you be able to have a look at my comments?

File: ibm/mas_devops/playbooks/oneclick_add_manage.yml [hunk 4]
```diff
@@ -60,6 +73,19 @@
           - lookup('env', 'MAS_JDBC_CERT_LOCAL_FILE') != ""
         fail_msg: "JDBC SSL certificate file name is not defined for external database"
 
+    - include_role:
+      # Cloud Pak for Data Platform (~1 1/2 hours)
+        name: ibm.mas_devops.cp4d
+      when: install_cp4d_platform == true
+      vars:
+        cpd_service_name: ca
```
[the remark this answers is not in the document]
Comment (lokesh-sreedhara) [reply ibm/mas_devops/playbooks/oneclick_add_manage.yml:81]: From our conversations, i remember This was to install cp4d base and spss as part of `Manage` playbook. if the user sets the env to true

File: docs/playbooks/oneclick-cognos.md [hunk 5]
```diff
@@ -0,0 +1,72 @@
+# Install Cognos Analytics Application
```
[the remark this answers is not in the document]
Comment (briangar-ibm) [reply docs/playbooks/oneclick-cognos.md:1]: These are explicit instructions to install Cognos.  No Cognos one click exists, if you read through the instructions it is to run the cp4d role.   Cognos install is enabled through the setting of environment variables.  It gives a blueprint for how a customer would install Cognos even after they have already installed MAS/Manage/CP4D.

File: ibm/mas_devops/playbooks/oneclick_add_manage.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (briangar-ibm) [reply ibm/mas_devops/playbooks/oneclick_add_manage.yml:81]: That is correct.  It gives the user the option to install CP4D and Cognos in the one_click manage installation.

File: ibm/mas_devops/roles/cp4d_service/defaults/main.yml (hunk 1 above)
Comment (briangar-ibm) [reply ibm/mas_devops/roles/cp4d_service/defaults/main.yml:49]: All 3 of these settings were required for the CAService to be installed properly

File: ibm/mas_devops/roles/cp4d_service/tasks/install.yml [hunk 6]
```diff
@@ -132,8 +133,8 @@
     namespace: "{{ cpd_operators_namespace }}"
     name: "cpd-{{ cpd_service_name }}-operator"
   register: cpd_sub_info
-  retries: 20
-  delay: 60 # Retry for approx 20 minutes (60s * 20 attempts) before giving up
+  retries: 40
```
[the remark this answers is not in the document]
Comment (briangar-ibm) [reply ibm/mas_devops/roles/cp4d_service/tasks/install.yml:136]: CCS including rabbitmq, elasticsearch and couchdb required a very long time to install.  This extra 20 minutes was to allow for those installations to occur.  Perhaps this was a connection/communication delay issue during my installation testing, but if I needed it, perhaps the user will also.

File: ibm/mas_devops/roles/cp4d/templates/subscriptions.yml.j2 [hunk 7]
```diff
@@ -8,7 +8,7 @@ spec:
   channel: "{{ cpd_channel }}"
   installPlanApproval: Automatic
   name: cpd-platform-operator
-  source: ibm-operator-catalog
+  source: {{ common_services_catalog_source }}
```
[the remark this answers is not in the document]
Comment (briangar-ibm) [reply ibm/mas_devops/roles/cp4d/templates/subscriptions.yml.j2:11]: I've made this change.

File: ibm/mas_devops/roles/cp4d/templates/subscriptions.yml.j2 [hunk 8]
```diff
@@ -20,5 +20,5 @@ spec:
   channel: {{ common_services_channel }}
   installPlanApproval: Automatic
   name: ibm-namespace-scope-operator
-  source: ibm-operator-catalog
+  source: {{ common_services_catalog_source }}
```
[the remark this answers is not in the document]
Comment (briangar-ibm) [reply ibm/mas_devops/roles/cp4d/templates/subscriptions.yml.j2:23]: I've made this change.

File: ibm/mas_devops/roles/cp4d_service/templates/services/ca.yml.j2 (hunk 3 above)
Comment (briangar-ibm) [reply ibm/mas_devops/roles/cp4d_service/templates/services/ca.yml.j2:8]: @andrercm Let's discuss this change in more detail.

File: docs/playbooks/oneclick-cognos.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (briangar-ibm) [reply docs/playbooks/oneclick-cognos.md:1]: changed oneclick-cognos.md to cp4d.md and made additional changes to document based on discussion with @andrercm

File: ibm/mas_devops/roles/cp4d_service/templates/services/ca.yml.j2 (hunk 3 above)
Comment (briangar-ibm) [reply ibm/mas_devops/roles/cp4d_service/templates/services/ca.yml.j2:8]: Removed fileStorageClass based on discussion with @andrercm

File: ibm/mas_devops/roles/cp4d_service/tasks/install.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (briangar-ibm) [reply ibm/mas_devops/roles/cp4d_service/tasks/install.yml:136]: Reverted back to 20 iterations

**Pull request merged**